Sync up to changes from Midwest FurFest

kitsuta opened the pull request:
I've been doing a bunch of development work on this branch, and it's about time to pull some of these fixes into MAGFest master.

Stuff in here that isn't very useful/relevant to MAGFest:
1. A new BADGE_TYPE_PRICES set of config options that allows events to have custom attendee-facing badge types with set prices.
2. Major improvements to the 'full address' collection UI, particularly, normalization of data.
3. Some bug fixes that don't apply to MAGFest's config settings.

Stuff in here that is helpful/relevant to MAGFest:
1. Dealer interface improvements, including the ability for dealers to edit their applications after applying. Note: the Marketplace head gets emailed whenever an application is updated.
2. UI improvements for the 'legal name' field.
3. A "refunded badge" page for badges with the refunded status, which lets attendees use the information from their old badge to pre-fill in the pre-reg form.
4. Stops redirecting group members to the group management page (since we now have the dealer form there, and we always considered stopping that behavior anyway)
5. Various bug fixes and UI improvements:
   1. https://github.com/magfest/ubersystem/commit/02b1f7bf1d222b45644d957fcce80dc77510da39
   2. https://github.com/magfest/ubersystem/commit/6d746e10b6a63ec97aa044356c84c637b283ebb5 (although we may already have this in MAGFest master)
   3. https://github.com/magfest/ubersystem/commit/c8b7fad252a1db1687bab0e10aed828fdb02b7ad
   4. https://github.com/magfest/ubersystem/commit/409b59df05dcbb06f007cc6800947d20310fa60b
   5. https://github.com/magfest/ubersystem/commit/5110a8a29906f4687a01574d4af50595f54687bf
   6. https://github.com/magfest/ubersystem/commit/b859014c37a545666bd89b62eaedc0a4e9a3e4c8
   7. https://github.com/magfest/ubersystem/commit/43ce109ff15517e74cb4aa40ada68a6cdc54a2f5 (plus https://github.com/magfest/ubersystem/commit/689527edd64264a4afecc57beaf98e8cc706bcbe)
   8. https://github.com/magfest/ubersystem/commit/f4556c1e9078e8b8d7e42d75efd753c6ade949c8
   9. https://github.com/magfest/ubersystem/commit/e09a31063fac34358ccb50abec35f97b1c9ebfe3

kitsuta commented:
Whoops, this will need some pep8 and unit test updates.

RobRuana commented:
Looks like this pull request is a work in progress? I haven't looked through it at all, so let me know when you'd like me to review it (even if it's just an incremental review or just to look at one thing).

Also, thanks for all the work! :heart:

kitsuta commented:
Yeah, there's still work to be done (since we don't have the tests running on the other fork, I didn't realize I was breaking stuff). I'll let you know when it's fully ready, but I wanted to compile a list of what I'd done so far so people can say "yes that's fine" or "no, this is a terrible idea."

As I mentioned in Slack, the more important thing to check is all the bugfixes -- these are almost certainly going to come up soon for MAGFest (or they already have), so it'd be good to look them over and/or cherry-pick them in so we don't repeat a ton of work.

kitsuta commented:
@RobRuana This branch is almost ready, just gotta add some unit tests.

kitsuta commented:
Huh, that's super interesting.

I added several unit tests but was confused when config options that weren't present in configspec.ini or test-defaults.ini seemed to exist. I added monkeypatching for them and committed that.

Those unit tests failed on the server because I was monkeypatching config options that didn't exist. It turns out that my tests were using my development.ini because I didn't have a test.ini.

I'm not sure the best way to solve this besides 'tell everyone to make a `test.ini` file'.

RobRuana commented:
Try adding the config options in test-defaults.ini.

RobRuana commented:
Does this pull request have all the StripeTransaction stuff as well? Is there any way we can keep those two issues separate?

kitsuta commented:
Oh whoops, it does, sorry! That slipped my mind. :(

I don't exactly see how to do that, except to just wait until we merge the StripeTransaction PRs. This branch in particular is what their staging server runs off, so I don't know how we'd handle that.

In general this process was fairly messy and I'd like to come up with a way to do this sort of work without creating these kinds of PRs or even ideally without having this kind of temporary split of the codebases. I don't really have good ideas for that, though.

RobRuana commented:
The easiest way is to continually make small merges back upstream. There's also the `git rerere` command, which I've never used and appears to be black magic.

RobRuana commented on uber/api.py:
```diff
@@ -103,4 +97,8 @@ def lookup(self, field):
+    services.register(AttendeeLookup(), 'attendee')
+    services.register(JobLookup(), 'shifts')
+    services.register(DepartmentLookup(), 'dept')
+    services.register(ConfigLookup(), 'config')
```
This is a great change!

RobRuana commented on uber/configspec.ini:
```diff
@@ -51,6 +51,11 @@ hostname = string(default="localhost")
+# We expose some basic services, like an attendee lookup, via WebSocket.
+# However, without a client cert configured, anyone can access these,
+# so this is off by default.
+api_enabled = boolean(default=False)
```
👍

kitsuta replied:
Amusingly, this was the cause of our one merge conflict. Who knew that both you and I would select the same spot in configspec.ini to add something?

kitsuta commented:
Because this is such a big PR with a lot of different features, I'm going to preserve the commit history and avoid squash+merge. In the future I'll endeavor to merge changes upstream (downstream? cross-stream?) a little more piecemeal.
